### ISSUE TITLE
BIP 44 Keychain support

### ISSUE THREAD

**Issue by msgilligan** [bitcoinj/bitcoinj#1869]
The `DeterministicKeyChain` class JavaDoc says that it implements BIP 32. I've been using it with BIP 44 paths and it seems to be working correctly, but the class could be modified/extended to better support BIP 44 (or possibly a subclass could be created) as there are some missing constants and behaviors that could be built-in.

**Comment by schildbach** [comment]
I'm not sure if `DeterministicKeyChain` is the right place to implement BIP44/49/84. It should work on any path. The only assumption is the internal/external separation of addresses which is currently hardcoded. And the lookahead.

Maybe we should move the `*_PATH` constants (but not the `*_SUBPATH` constants) to `KeyChainGroup` or to `KeyChainGroupStructure`?

I'm also not sure if we _need_ to support BIP49 – I'd say we _could_. Personally I hope we can skip it. P2WPKH is the way to go and P2PKH works as a fallback.

Regarding BIP84 … we could start by providing a `KeyChainGroupStructure` implementation that uses BIP44 for P2PKH and BIP84 for P2WPKH.

**Comment by msgilligan** [comment]
> I'm not sure if DeterministicKeyChain is the right place to implement BIP44/49/84. The class should work on any path.

Agreed.

> Maybe we should move the *_PATH constants (but not the *_SUBPATH constants) to KeyChainGroup or to KeyChainGroupStructure?

Yes, that sounds about right.

> I'm also not sure if we need to support BIP49 – I'd say we could. Personally I hope we can skip it. P2WPKH is the way to go and P2PKH works as a fallback.

Yeah, I don't think it is very important (and I don't think I need it for anything I'm planning on doing) but we should probably provide the "purpose" constant and make sure we don't preclude anyone from using BIP 49 if they need to.

> Regarding BIP84 … we could start by providing a KeyChainGroupStructure implementation that uses BIP44 for P2PKH and BIP84 for P2WPKH?

That sounds like the right approach. I also have some ideas and code for an `HDPath` class that replaces (but is compatible with) the current use of `ImmutableList<ChildNumber>` that should make the code a little more readable and maintainable.

**Comment by msgilligan** [comment]
PR #1877 adds the `HDPath` class which might be a good place to put some of the `*_PATH` constants.

**Comment by HashEngineering** [comment]
The `DeterministicKeyChain` class does work on other paths.  Over 1 year ago, such support was added.   In a fork of this project I create wallets with BIP32 and BIP44 paths to keep compatibility with wallets that only have BIP32.

Recently, I have been looking at bitcoinj 0.15 and found that the `KeyChainGroupStructure` class is interesting and a promising start to addressing this issue.

**Comment by msgilligan** [comment]
> the `KeyChainGroupStructure` class is interesting and a promising start to addressing this issue.

Yes, I have created an implementation of that interface in another project that I plan on using as the basis of a future PR.

**Comment by msgilligan** [comment]
OK, PR #2297 is ready to look at.

**Comment by msgilligan** [comment]
PR #2297 is merged.

Currently being reviewed:
PR #2299 adds BIP-43 structure support to `WalletTemplate`.
PR #2437 adds BIP-43 structure support to `wallet-tool`.

This comment has my thoughts about a roadmap for future BIP-43 related work: https://github.com/bitcoinj/bitcoinj/pull/2299#issuecomment-1113883596
Maybe that comment should be copied here as we continue on this path.

**Comment by msgilligan** [comment]
Note: This document is relevant and we should probably read and update it: https://github.com/bitcoinj/bitcoinj/blob/master/designdocs/Deterministic%20wallets.txt

**Comment by schildbach** [comment]
> Note: This document is relevant and we should probably read and update it: https://github.com/bitcoinj/bitcoinj/blob/master/designdocs/Deterministic%20wallets.txt

I'll try my luck.

**Comment by schildbach** [comment]
See https://github.com/bitcoinj/bitcoinj/pull/2450
